File: frontend/mobile_app/lib/data/customer_repository.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../core/config/app_config.dart';
import '../models/customer.dart';

/// Customer onboarding data access: the option lists and customer capture.
/// Swapped for [MockCustomerRepository] in standalone mode.
///
/// [create] persists the full (all-optional) profile once; [update] applies a
/// partial change to an existing record — so the associate can fill parts of the
/// form across multiple saves (sizes now, colours later) without losing anything.
abstract interface class CustomerRepository {
  Future<OnboardingOptions> options();

  Future<Customer?> create(Customer draft, {String? sessionId});

  Future<Customer?> update(Customer draft, {String? sessionId});
}

/// [CustomerRepository] backed by the Node backend (`/api/customers*`).
class HttpCustomerRepository implements CustomerRepository {
  HttpCustomerRepository({http.Client? client})
    : _client = client ?? http.Client();

  final http.Client _client;
  static const Duration _timeout = Duration(seconds: 8);

  @override
  Future<OnboardingOptions> options() async {
    try {
      final http.Response res = await _client
          .get(AppConfig.http('/api/customers/options'))
          .timeout(_timeout);
      if (res.statusCode != 200) return OnboardingOptions.empty;
      return OnboardingOptions.fromJson(
        jsonDecode(res.body) as Map<String, dynamic>,
      );
    } catch (_) {
      return OnboardingOptions.empty;
    }
  }

  /// POST the full profile (partial bodies are fine — every field is optional).
  @override
  Future<Customer?> create(Customer draft, {String? sessionId}) async {
    final Map<String, dynamic> body = _body(draft, sessionId);
    final http.Response res = await _client
        .post(
          AppConfig.http('/api/customers'),
          headers: _jsonHeaders,
          body: jsonEncode(body),
        )
        .timeout(_timeout);
    if (res.statusCode != 200 && res.statusCode != 201) {
      throw Exception('Failed to save customer (${res.statusCode})');
    }
    return Customer.fromJson(jsonDecode(res.body) as Map<String, dynamic>);
  }

  /// PUT a partial update onto an existing record. Only the fields present on
  /// [draft] change server-side; untouched fields are preserved.
  @override
  Future<Customer?> update(Customer draft, {String? sessionId}) async {
    if (draft.id.isEmpty || draft.id == 'draft') {
      // No server id yet — fall back to a create so nothing is lost.
      return create(draft, sessionId: sessionId);
    }
    final Map<String, dynamic> body = _body(draft, sessionId);
    final http.Response res = await _client
        .put(
          AppConfig.http('/api/customers/${draft.id}'),
          headers: _jsonHeaders,
          body: jsonEncode(body),
        )
        .timeout(_timeout);
    if (res.statusCode != 200) {
      throw Exception('Failed to update customer (${res.statusCode})');
    }
    return Customer.fromJson(jsonDecode(res.body) as Map<String, dynamic>);
  }

  static const Map<String, String> _jsonHeaders = <String, String>{
    'content-type': 'application/json',
  };

  /// The customer's partial JSON plus an optional sessionId link. [Customer.toJson]
  /// already omits null/empty fields, giving us a clean partial body.
  Map<String, dynamic> _body(Customer draft, String? sessionId) {
    final Map<String, dynamic> body = draft.toJson();
    if (sessionId != null && sessionId.isNotEmpty) body['sessionId'] = sessionId;
    return body;
  }
}

/// Offline/demo customer capture: local option lists, synthetic customer id.
class MockCustomerRepository implements CustomerRepository {
  const MockCustomerRepository();

  @override
  Future<OnboardingOptions> options() async => const OnboardingOptions(
    genders: <String>['Female', 'Male', 'Non-binary', 'Prefer not to say'],
    ageRanges: <String>['Under 18', '18-24', '25-34', '35-44', '45-54', '55+'],
    personalities: <String>[
      'Classic',
      'Minimalist',
      'Trendsetter',
      'Bold',
      'Romantic',
      'Sporty',
    ],
  );

  @override
  Future<Customer?> create(Customer draft, {String? sessionId}) async =>
      draft.id.isEmpty || draft.id == 'draft'
      ? draft.copyWith(id: 'mock_customer')
      : draft;

  @override
  Future<Customer?> update(Customer draft, {String? sessionId}) async => draft;
}

File: frontend-rn/mobile_app/src/data/customerRepository.ts
import { AppConfig } from '../config/appConfig';
import { Customer, OnboardingOptions } from '../models/customer';

/// Customer onboarding data access: option lists + customer capture. Ported 1:1
/// from the Flutter mobile `CustomerRepository`.
export interface CustomerRepository {
  options(): Promise<OnboardingOptions>;
  create(draft: Customer, opts?: { sessionId?: string }): Promise<Customer | undefined>;
  update(draft: Customer, opts?: { sessionId?: string }): Promise<Customer | undefined>;
}

export class HttpCustomerRepository implements CustomerRepository {
  private static readonly timeoutMs = 8000;

  private async fetchTimeout(url: string, init?: RequestInit): Promise<Response> {
    const controller = new AbortController();
    const t = setTimeout(() => controller.abort(), HttpCustomerRepository.timeoutMs);
    try {
      return await fetch(url, { ...init, signal: controller.signal });
    } finally {
      clearTimeout(t);
    }
  }

  async options(): Promise<OnboardingOptions> {
    try {
      const res = await this.fetchTimeout(AppConfig.http('/api/customers/options'));
      if (res.status !== 200) return OnboardingOptions.empty;
      return OnboardingOptions.fromJson(await res.json());
    } catch {
      return OnboardingOptions.empty;
    }
  }

  async create(draft: Customer, opts?: { sessionId?: string }): Promise<Customer | undefined> {
    const res = await this.fetchTimeout(AppConfig.http('/api/customers'), {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(this.body(draft, opts?.sessionId)),
    });
    if (res.status !== 200 && res.status !== 201) {
      throw new Error(`Failed to save customer (${res.status})`);
    }
    return Customer.fromJson(await res.json());
  }

  async update(draft: Customer, opts?: { sessionId?: string }): Promise<Customer | undefined> {
    if (draft.id.length === 0 || draft.id === 'draft') {
      // No server id yet — fall back to a create so nothing is lost.
      return this.create(draft, opts);
    }
    const res = await this.fetchTimeout(AppConfig.http(`/api/customers/${draft.id}`), {
      method: 'PUT',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(this.body(draft, opts?.sessionId)),
    });
    if (res.status !== 200) throw new Error(`Failed to update customer (${res.status})`);
    return Customer.fromJson(await res.json());
  }

  private body(draft: Customer, sessionId?: string): Record<string, any> {
    const body = draft.toJson();
    if (sessionId != null && sessionId.length > 0) body.sessionId = sessionId;
    return body;
  }
}

/// Offline/demo customer capture.
export class MockCustomerRepository implements CustomerRepository {
  async options(): Promise<OnboardingOptions> {
    return new OnboardingOptions({
      genders: ['Female', 'Male', 'Non-binary', 'Prefer not to say'],
      ageRanges: ['Under 18', '18-24', '25-34', '35-44', '45-54', '55+'],
      personalities: ['Classic', 'Minimalist', 'Trendsetter', 'Bold', 'Romantic', 'Sporty'],
    });
  }

  async create(draft: Customer): Promise<Customer | undefined> {
    return draft.id.length === 0 || draft.id === 'draft' ? draft.copyWith({ id: 'mock_customer' }) : draft;
  }

  async update(draft: Customer): Promise<Customer | undefined> {
    return draft;
  }
}
